### PR TITLE
feat: Add distributed version of `ChunkSampler`

### PR DESCRIPTION
**Pull request by felix0097**
This PR adds an extension of the `ChunkSampler` class that shards the dataset for training using `torch.distributed`.

### CODE REVIEW

**Review by ilan-gold — COMMENTED** [review]
This should be made generlizable.  We can make a `IsDistributable` `Protocol` that the samplers inherit.  That way anything that has implements a `_mask` attribute can be handled this way.

To do this, we may need to consider making `_mask` a bit more general, like a `list[slice]` for things like categorical samplers

Maybe I'm misunderstanding though, it seems like this is just masking.

Also no reason to have a class-per-package - if everything is just based on rank/world, we can make this generic (i.e., users can pass in rank/world if `torch` or one of the recognized packages is not installed)

**Review by ilan-gold — COMMENTED** [review]
If @giovp approves, so do I (same for @selmanozleyen )

File: tests/test_sampler.py [hunk 1]
```diff
@@ -369,3 +370,172 @@ def test_automatic_batching_respects_shuffle_flag(shuffle: bool):
...
+                world_size=world_size,
+                chunk_size=chunk_size,
+                preload_nchunks=preload_nchunks,
+                batch_size=batch_size,
+                enforce_equal_batches=True,
+            )
+            n_batches = sum(len(lr["splits"]) for lr in sampler.sample(n_obs))
+            batch_counts.append(n_batches)
+
+        assert len(set(batch_counts)) == 1, f"Batch counts differ across ranks: {batch_counts}"
```
Comment (selmanozleyen) [inline tests/test_sampler.py]: why not assert this to batch_size also like batch_counts[0] ==  batch_size

File: tests/test_sampler.py [hunk 2]
```diff
@@ -369,3 +370,172 @@ def test_automatic_batching_respects_shuffle_flag(shuffle: bool):
...
+        sampler = make_distributed_sampler(
+            rank=0,
+            world_size=world_size,
+            chunk_size=chunk_size,
+            preload_nchunks=preload_nchunks,
+            batch_size=batch_size,
+            enforce_equal_batches=True,
+        )
+        indices = collect_indices(sampler, n_obs)
+        assert len(set(indices)) == 30
```
Comment (selmanozleyen) [inline tests/test_sampler.py]: is this different when len(indices)? from what I know we shouldn't have overlapping indices right?

**Comment by selmanozleyen** [comment]
>  @selmanozleyen how would distributed training over a collection that is shuffled-by-category look? Probably masking out subsets of each category? But also giving categories per node? This probably depends on how the weights are merged based on my limited understanding.
 
We don't need to share categories per nodes unless we have a concept of epoch (sampling without replacement). In fact we can't share categories per node  during whole training if `num_categories %  num_nodes != 0` and weights would make it more complicated. So I wouldn't support sampling without replacement on distributed training for categoricals.

File: src/annbatch/samplers/_chunk_sampler.py [hunk 3]
```diff
@@ -184,3 +184,128 @@ def _compute_chunks(self, chunk_indices: np.ndarray, start: int, stop: int) -> l
         offsets = np.cumsum(offsets)
         starts, stops = offsets[:-1][chunk_indices], offsets[1:][chunk_indices]
         return [slice(int(s), int(e)) for s, e in zip(starts, stops, strict=True)]
+
+
+def _get_dist_info_torch() -> tuple[int, int]:
+    """Get rank and world_size from ``torch.distributed``."""
```
[the remark this answers is not in the document]
Comment (felix0097) [reply src/annbatch/samplers/_chunk_sampler.py]: Fair point, I will add the option to provide a callable as well. Callable then just return a tuple `(rank, world_size)` then people can easily extend this to other frameworks. 

I would keep the torch + jax option though. Most people who do distributed training are not really aware of those concepts. E.g lightning [abstracts those away](https://lightning.ai/docs/pytorch/stable/accelerators/gpu_intermediate.html). So, it's easy for them to use it, they just need to know which framework their using.

**Comment by ilan-gold** [comment]
## API

It sounds like there would not be much overlap then so my original comment 

>  As a V1 this seems fine to me, we can always extend/grow the requirements ...

holds @felix0097 so I think leaving the usage of `_mask` is fine for now. If we run across another use-case, we can generalize, but it sounds like the requirements for categorical samplers would be different.

## RNG

> Maybe "bad for training" is too strong since it is hard to prove with concrete examples but in theory if you write down the training, batches are sampled with i.i.d. assumptions. Having same batch indices would hurt the assumptions more

But if the data is pre-shuffled, we wouldn't care about having the same batch indices, no?

**Comment by felix0097** [comment]
Not sure I fully understand your statement @selmanozleyen. In general, we always load several chunks at the same time (e.g 128) and shuffle them in memory. So we do break up the on disk chunks structure that the model can't simply "remember" on disk chunks

**Review by selmanozleyen — APPROVED** [review]
Lgtm.

**Review by ilan-gold — COMMENTED** [review]
Ok last question - I know this would be a big re-architecture (eh, not that big) but something we accidentally have as a result of `ChunkSampler.mask` as an arg is training-test-validation splits with shuffled datasets.

How would one do that here? I think this is a point in favor of baking this into `ChunkSampler`
